lib/faraday/response.rb

```diff
@@ -61,8 +61,8 @@ def on_complete
 
     def finish(env)
       raise "response already finished" if finished?
-      @on_complete_callbacks.each { |callback| callback.call(env) }
-      @env = Env.from(env)
+      @env = env.is_a?(Env) ? env : Env.from(env)
```

Comment:
I think it's a feature that `Env.from` creates a copy by itself. In this case, we don't want a copy so we put up a guard. That's the reason I didn't cherry pick your original commit.